### PR TITLE
add worst_tiles sampling

### PR DESCRIPTION
**Pull request by JonasHell**
Current implementation looks quite complicated. Unfortunately, the idea is not compatible with your `_score_based_points()`. But maybe there still some room to improve my implementation...

### CODE REVIEW

File: torch_em/shallow2deep/prepare_shallow2deep.py [hunk 1]
```diff
@@ -504,11 +509,94 @@ def random_points(
...
+    onehot = np.eye(unique.shape[0])[inverse]
+
+    # compute the difference between labels and prediction
+    diff = np.abs(onehot - pred).sum(axis=1)
+    assert len(diff) == len(features)
+
+    # reshape diff to image shape and apply convolution with 1-kernel
+    diff_img = torch.Tensor(diff.reshape(img_shape)[None, None])
+    kernel = torch.Tensor(np.ones(tiles_shape)[None, None])
+    diff_img_smooth = torch.nn.functional.conv2d(diff_img, weight=kernel) if len(img_shape)==2 \
```
Comment (constantinpape) [inline torch_em/shallow2deep/prepare_shallow2deep.py:541]: I think using torch for the convolution here is a bit overkill. Two points:
- Do we actually want to apply a uniform kernel here, or do we rather want to apply smoothing with a large sigma (for this can use `vigra.filters.gaussianSmoothing`)
- If we do want the uniform kernel you can use the convolve functionality from scipy, speed doesn't matter here and it's easier to read because you don't need to convert to torch tensors: `scipy.ndimage.convolve`.

File: torch_em/shallow2deep/prepare_shallow2deep.py [hunk 2]
```diff
@@ -504,11 +509,94 @@ def random_points(
...
+    # of the label-prediction diff
+    nc = len(np.unique(labels))
+    n_samples_class = int(sample_fraction_per_stage * len(features)) // nc
+
+    # sample in a class balanced way
+    samples_per_class = [[]]*nc
+    indices_per_class = [np.where(labels == class_id)[0] for class_id in range(nc)]
+
+    # get maxima of the label-prediction diff
+    max_centers = np.argsort(diff_img_smooth)[::-1]
```
Comment (constantinpape) [inline torch_em/shallow2deep/prepare_shallow2deep.py:562]: Wouldn't we rather use a function that finds local maxima here? I think that would make more sense, because otherwise we can end up oversampling the global maximum.
There's https://scikit-image.org/docs/dev/api/skimage.feature.html#skimage.feature.peak_local_max or `vigra.analysis.localMaxima` you can use for this.

**Review by constantinpape — COMMENTED** [review]
The idea looks very good!
I left a couple of comments of how I think we could improve it.

**Comment by constantinpape** [comment]
Oh, and just disregard the CI failures, this looks like a conda issue; hopefully it's transient and things will work again tomorrow.

File: torch_em/shallow2deep/prepare_shallow2deep.py [hunk 3]
```diff
@@ -429,6 +431,7 @@ def worst_points(
     forests, forests_per_stage,
     sample_fraction_per_stage,
     accumulate_samples=True,
+    sampling_kwargs={},
```
[the remark this answers is not in the document]
Comment (JonasHell) [reply torch_em/shallow2deep/prepare_shallow2deep.py:434]: I thought in this case I have to adapt the arguements of all the other sampling startegies as well, but when I think about it again right now that's not true. So sure, I'll fix it.

File: torch_em/shallow2deep/prepare_shallow2deep.py (hunk 1 above)
Comment (JonasHell) [reply torch_em/shallow2deep/prepare_shallow2deep.py:541]: Sure, I'll replace torch. About the question what's better, I don't know to be honest. What do you think? Maybe it's possible to allow the smoothing function to be given as an input. Or do you think that's a bit overkill?

File: torch_em/shallow2deep/prepare_shallow2deep.py (hunk 2 above)
Comment (JonasHell) [reply torch_em/shallow2deep/prepare_shallow2deep.py:562]: Yeah, you're totally right. I could even add the `min_distance` parameter. Then I could make sure that there is no overlap of the tiles at all. Maybe then I can do the entire thing more efficient. I'll have a look into that :)

File: torch_em/shallow2deep/prepare_shallow2deep.py (hunk 1 above)
Comment (constantinpape) [reply torch_em/shallow2deep/prepare_shallow2deep.py:541]: > What do you think? Maybe it's possible to allow the smoothing function to be given as an input. Or do you think that's a bit overkill?

Maybe we can keep it a bit simpler and have a single parameter, that is either a scalar (then we interpret it as sigma and use gaussian smoothing) or a tuple (then we interpret it as tile shape and apply the uniform kernel)

**Comment by JonasHell** [comment]
I adapted all your suggestion and additionally made the code faster and easier to read.
In the current version I removed the early stopping. While trying a few things it didn't feel like it brings any speed boost. Rather the opposite was the case, it felt like the overhead of the if statements made it slower. But maybe this impression is wrong... What do you think: better keep the current easy to read version with 2 seperate for loops (where we iterate over all maxima even if we don't need all) or add the nested for loop with early stopping back in?

**Comment by constantinpape** [comment]
> I adapted all your suggestion and additionally made the code faster and easier to read.

Yes, looks much better now.

> better keep the current easy to read version with 2 seperate for loops 

Yes, let's keep the current version, it reads much easier and we don't really care about speed here.


What I am not quite sure about yet: `[51, 51]` sounds like a fairly large shape to take around the maxima. How did you choose it?

File: torch_em/shallow2deep/prepare_shallow2deep.py [hunk 4]
```diff
@@ -504,11 +507,86 @@ def random_points(
     return features[samples], labels[samples]
 
 
+def worst_tiles(
+    features, labels, rf_id,
+    forests, forests_per_stage,
+    sample_fraction_per_stage,
+    img_shape,
+    tiles_shape=[51, 51],
```
[the remark this answers is not in the document]
Comment (JonasHell) [reply torch_em/shallow2deep/prepare_shallow2deep.py:515]: To be honest, I don't have a real good metric for that. Just tried with 7, 25, and 51 and the DiceLoss was smallest for 51. But you already mentioned that that's not a good proxy. So I guess we can use something smaller. What would you suggest?
With 51 usually 2-3 maxima were sampled :)

[the remark this answers is not in the document]
Comment (JonasHell) [reply torch_em/shallow2deep/prepare_shallow2deep.py:515]: with a patch size between 256 and 512

[the remark this answers is not in the document]
Comment (constantinpape) [reply torch_em/shallow2deep/prepare_shallow2deep.py:515]: > With 51 usually 2-3 maxima were sampled :)

But did it actually then sample points from all of them after class balancing?

[the remark this answers is not in the document]
Comment (constantinpape) [reply torch_em/shallow2deep/prepare_shallow2deep.py:515]: Also, now that I think about the class balancing more here: would it maybe make sense to compute the maxima for the classes independently? I think we might run into cases where we only sample one class otherwise.

[the remark this answers is not in the document]
Comment (JonasHell) [reply torch_em/shallow2deep/prepare_shallow2deep.py:515]: > > With 51 usually 2-3 maxima were sampled :)
> 
> But did it actually then sample points from all of them after class balancing?

at least one class did sample from all of them, but you are right for the others that's not guaranteed

[the remark this answers is not in the document]
Comment (JonasHell) [reply torch_em/shallow2deep/prepare_shallow2deep.py:515]: > Also, now that I think about the class balancing more here: would it maybe make sense to compute the maxima for the classes independently? I think we might run into cases where we only sample one class otherwise.

True, this could happen. I'll try to adapt it :)

[the remark this answers is not in the document]
Comment (JonasHell) [reply torch_em/shallow2deep/prepare_shallow2deep.py:515]: but then we loose a bit of the "a human would label localy" idea. because locality is then just guaranteed in a class but not over classes. But I think this still makes sense.

**Comment by JonasHell** [comment]
Finding the maxima is now done per class.
I also changed default tile size to 25. Would you prefer it to be even smaller?

**Review by constantinpape — APPROVED** [review]
Thanks, I think the changes make sense now, and the per class maxima should ensure a more balanced sampling of classes; and 25 sounds like a good default.
